### PR TITLE
feat: Add EcPairing precompile

### PR DESCRIPTION
**Pull request by jrchatruc**
# What ❔

This PR adds the EcPairing precompile from https://github.com/lambdaclass/zksync_era_precompiles/, commit [4bdfebf831e21d58c5ba6945d4524763f1ef64d4](https://github.com/lambdaclass/zksync_era_precompiles/commit/4bdfebf831e21d58c5ba6945d4524763f1ef64d4).

Part of EVM-572

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: system-contracts/contracts/precompiles/EcPairing.yul [hunk 1]
```diff
@@ -0,0 +1,1743 @@
...
+                ret := iszero(or(x, y))
+            }
+
+            /// @notice Checks if a point in affine coordinates in Montgomery form is on the curve.
+            /// @dev The curve in question is the alt_bn128 curve.
+            /// @dev The Short Weierstrass equation of the curve is y^2 = x^3 + 3.
+            /// @param x The x coordinate of the point in Montgomery form.
+            /// @param y The y coordinate of the point in Montgomery form.
+            /// @return ret True if the point is on the curve, false otherwise.
+			function g1AffinePointIsOnCurve(x, y) -> ret {
```
Comment (StanislavBreadless) [inline system-contracts/contracts/precompiles/EcPairing.yul:377]: probably mix of tabs/spaces, it looks weird in the github PR diff
image

File: system-contracts/SystemContractsHashes.json [hunk 2]
```diff
@@ -3,120 +3,127 @@
     "contractName": "AccountCodeStorage",
     "bytecodePath": "artifacts-zk/contracts-preprocessed/AccountCodeStorage.sol/AccountCodeStorage.json",
     "sourceCodePath": "contracts-preprocessed/AccountCodeStorage.sol",
-    "bytecodeHash": "0x01000075e8397f3d02eddfe166a8324a2ac31a6da0db8eeebfa8739d3f31aad6",
-    "sourceCodeHash": "0xfbf66e830201c4b7fda14f0ddf28a53beb7fbb48a8406392bcfd0ef7ea9265c8"
+    "bytecodeHash": "0x01000081df25f20524a83dc78040b54bdd50e97de2381e220757f07a56a9dd8b",
```
Comment (mm-zk) [inline system-contracts/SystemContractsHashes.json:6]: why are there all these changes? did you use a different compiler version?

**Comment by IAvecilla** [comment]
The table below presents the gas usage reports for the `ecpairing` precompile, comparing its performance before and after all the optimizations suggested after the audit.

| Test case | Gas used with optimizations | Gas used without optimizations |
| --------- | -------- | ----|
| ecpairing_empty_data_insufficient_gas  | 65473 | 92281 |
| ecpairing_one_point_with_g1_zero  | 1631296 | 943352 |
| ecpairing_one_point_with_g2_zero  | 1410436 | 93092 |
| ecpairing_empty_data  | 65473 | 92281 |
| ecpairing_one_point_fail  | 3144003 |3923454 |
| ecpairing_one_point_insufficient_gas  | 3144003 | 3923454 |
| ecpairing_fuzz_negative_1 | 3364750 | 3922908|
| ecpairing_three_point_fail_1  | 6616353 | 11584084 |
| ecpairing_three_point_match_1  | 4880704 | 7753722|
| ecpairing_two_points_with_one_g2_zero  | 3144592 |3924055|
| ecpairing_two_point_fail_2  | 4879413 |7753925 |
| ecpairing_two_point_fail_1  | 4879539 |7754507 |
| ecpairing_two_point_match_1  | 4879611 |7754597 |
| ecpairing_two_point_match_2  | 4879611 |7754597 |
| ecpairing_two_point_match_5  | 1631885 |943953 |
| ecpairing_two_point_match_4  | 4880115 |7753121 |
| ecpairing_two_point_match_3  | 4879959 |7753409 |
| ecpairing_fuzz_negative_2  | 4879761 | 4773613|
| ecpairing_two_point_oog  | 4879611 |7754597 |
| ecpairing_fuzz_negative_3  | 3140000 | 3923000 |
| ecpairing_fuzz_negative_4  | 11824649 | 23077095|
| ecpairing_fuzz_positive_1  | 4879000 | 7753900|
| ecpairing_fuzz_positive_2 | 1411000 | 94300 |
| ecpairing_fuzz_positive_3 | 2074194 | 2645074 |
| ecpairing_fuzz_positive_4  | 8351577 | 15414267 |
| ecpairing_fuzz_positive_5  | 10087461 | 19244012|

File: system-contracts/contracts/precompiles/EcPairing.yul (hunk 1 above)
Comment (jrchatruc) [reply system-contracts/contracts/precompiles/EcPairing.yul]: Fixed in this commit [99af881](https://github.com/matter-labs/era-contracts/pull/383/commits/99af881ddda1886b7f374ddd17c3bc725233aa39)

**Pull request merged**